query/src/servers/clickhouse/interactive_worker_base.rs

```diff
@@ -92,22 +98,81 @@ impl InteractiveWorkerBase {
 
         let sc = sample_block.schema().clone();
         let stream = ReceiverStream::new(rec);
-        let stream = FromClickHouseBlockStream {
+        let ck_stream = FromClickHouseBlockStream {
             input: stream,
-            schema: sc,
+            schema: sc.clone(),
         };
 
         let interpreter = InterpreterFactory::get(ctx.clone(), PlanNode::Insert(insert))?;
         let name = interpreter.name().to_string();
 
+        if ctx.get_settings().get_enable_new_processor_framework()? != 0
+            && ctx.get_cluster().is_empty()
+        {
+            let output_port = OutputPort::create();
+            let sync_receiver_ck_source = SyncReceiverCkSource::create(
+                ctx.clone(),
+                ck_stream.input.into_inner(),
+                output_port.clone(),
+                sc,
+            )?;
+            let mut source_pipe_builder = SourcePipeBuilder::create();
+            source_pipe_builder.add_source(output_port, sync_receiver_ck_source);
+
+            // Get the specific `InterceptorInterpreter`, then the inner `InsertInterpreter`
+            let interceptor_interpreter = match interpreter
+                .as_any()
+                .downcast_ref::<InterceptorInterpreter>()
+            {
+                Some(interceptor) => interceptor,
+                None => panic!("Interpreter isn't a InterceptorInterpreter!"),
+            };
+            // In the context of this function, inner must be `InsertInterpreter`
+            let insert_interpreter = interceptor_interpreter.get_inner();
+            // Get the specific `InsertInterpreter`
+            let insert_interpreter = match insert_interpreter
+                .as_any()
+                .downcast_ref::<InsertInterpreter>()
+            {
+                Some(insert) => insert,
+                None => panic!("Interpreter isn't a InsertInterpreter!"),
+            };
+            let mut insert_interpreter_box = insert_interpreter.clone().get_box();
+            // Set `SourcePipeBuilder` to `InsertInterpreter`, used in insert source is `StreamingWithFormat`
+            insert_interpreter_box
+                .as_mut()
+                .set_source_pipe_builder(Some(source_pipe_builder));
+
+            // Set the newest `InsertInterpreter` to `InterceptorInterpreter`
+            let mut interceptor_interpreter_box = interceptor_interpreter.clone().get_box();
```

Comment:
This is too hacky. 

I think we can just add one method and make it simple.

```
pub fn set_source_pipe_builder(&self, builder: Option<SourcePipeBuilder>) -> Self {
        let mut res = self.clone();
        res.source_pipe_builder = builder; 
        res
    }
```

Reply:
So we need to add `set_source_pipe_builder` in `interpreter` trait?

Reply:
Yes,or we can make it:

default function:
```
pub fn set_source_pipe_builder(&self, builder: Option<SourcePipeBuilder>) {        
    }
```


insert/interpreter executor override:

```
pub fn set_source_pipe_builder(&self, builder: Option<SourcePipeBuilder>) {    
	    let mut lock = self. source_pipe_builder.write();  // RWMutext
       ....

    }
```

Reply:
Nice, will use the second way to make code simple!